keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,145 @@
+
+### Goals
+
+- Allow slow starting containers to run safely during startup with health probes enabled.
+- Improve documentation of the `Probe` structure in core types' API.
+- Improve `kubernetes.io/docs` section about Pod lifecycle:
+  - Clearly state that PostStart handlers do not delay probe executions.
+  - Introduce and explain this new option.
+  - Document that `kubelet` does not save states, and what are the implications with this new option (see risks and mitigations).
+  - Document appropriate use cases for this new option.
+
+### Non-Goals
+
+- Update [A Pod's Life] and other well referenced pages explaining probes (only API and official docs will be covered).
+
+[A Pod's Life]: https://blog.openshift.com/kubernetes-pods-life/
+
+## Proposal
+
+### Implementation Details
+
+The proposed solution is to add a new `int32` field named `MaxInitialFailureCount` to the type `Probe` of the core API, which is mapped to the numerical option `maxInitialFailureCount`.
+
+It also require keeping the state of the container (has the probe ever succeeded?) using a boolean `hasInitialized` inside the kubelet worker.
+
+The combination of `hasInitialized` false and `resultRun` count lower than `MaxInitialFailureCount` becomes another condition to return `true` to the probe state in `worker.go`.
+
+Adding a boolean `hasInitialized` to the kubelet worker keeping track of the state of the container (has the probe ever succeeded?) and use it as a new condition to return `true` to the probe state in `worker.go`.
+
+For example, if `periodSeconds` is 10, `maxInitialFailureCount` is 20, and `failureThreshold` is 3, it means that:
+
+- The kubelet will allow the container 200 seconds to start (20 probes, spaced 10 seconds apart).
+- If a probe succeeds at any time during that interval, the container is considered to have started, and `failureThreshold` is used thereafter.
+
+This means that all these cases will lead to a container terminated:
+
+- The container fails 20 probes at startup. It is considered to have failed, and is terminated after 200 seconds of downtime.
+- The container fails 10 probes at startup, starts successfully, and after a long time fails 3 probes. The container is considered to have failed and is terminated after 30 seconds of downtime.
+- The container fails 10 probes at startup, succeeds once, and fails 3 more probes. The container is considered to have started at 100 seconds, and even though it is still within the first 200 seconds of its lifetime covered by `maxInitialFailureCount`, it is considered to have failed because of `failureThreshold` and the fact that it had initially started successfully, therefore it is terminated after 30 seconds of downtime.
+
+You can see the implementation in PR [#71449].
+
+[#71449]: https://github.com/kubernetes/kubernetes/pull/71449
+
+### Risks and Mitigations
+
+#### Stateless kubelet
```

Comment:
This is really a limitation rather than a risk; you don't have a way at present of persisting `hasInitialized`.  But OK, it's reasonable enough to put it here.